### PR TITLE
HIVE-26659: Proceed MapJoin with empty storage if it performs AntiJoin

### PR DESCRIPTION
**Pull request by ngsg**
### What changes were proposed in this pull request?

Proceed MapJoin operation with empty storage if that storage is RHS table of AntiJoin.

### Why are the changes needed?

Current MapJoinOperator does not join empty RowContainers unless it has any OuterJoin.
Therefore, MapJoinOperator always produces no rows for AntiJoin as it joins non-empty LHS RowContainer and empty RHS RowContainer.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

I add a qfile test that does not use vectorized MapJoin operator.
I also ran all TPC-DS queries on 1TB dataset with hive.auto.convert.anti.join=true and checked that all queries returns the correct result.

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/MapJoinOperator.java [hunk 1]
```diff
@@ -556,6 +556,9 @@ public void process(Object row, int tag) throws HiveException {
               }
             } else {
               storage[pos] = emptyList;
+              if (pos != 0 && condn[pos - 1].getType() == JoinDesc.ANTI_JOIN) {
```
Comment (scarlin-cloudera) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/MapJoinOperator.java:559]: Can we add a comment here?  Thanks!

File: ql/src/test/queries/clientpositive/antijoin2.q [hunk 2]
```diff
@@ -0,0 +1,41 @@
+set hive.merge.nway.joins=false;
```
Comment (scarlin-cloudera) [inline ql/src/test/queries/clientpositive/antijoin2.q:1]: Can we also add the explain cbo plan and explain plan here?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/MapJoinOperator.java (hunk 1 above)
Comment (ngsg) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/MapJoinOperator.java:559]: Thank you for your review. I added a comment which explains the reason for proceeding join.

File: ql/src/test/queries/clientpositive/antijoin2.q (hunk 2 above)
Comment (ngsg) [reply ql/src/test/queries/clientpositive/antijoin2.q:1]: I added explain and explain cbo for each of the queries.

File: ql/src/test/queries/clientpositive/antijoin2.q [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+create table tt2 (ws_order_number bigint);
+create table tt3 (wr_order_number bigint);
+
+insert into tt1 values (42, 3093.96), (1041, 299.28), (1378, 85.56), (1378, 719.44), (1395, 145.68);
+insert into tt2 values (1378), (1395);
+insert into tt3 values (42), (1041);
+
+-- The result should be the same regardless of vectorization.
+
+explain
```
Comment (amansinha100) [inline ql/src/test/queries/clientpositive/antijoin2.q:20]: The plans for these queries don't show the pattern of MergeJoin --> MapJoin.  Without this pattern, as I noted in the Jira, the wrong results issue was not reproducible.  In order to force this plan, I had to set the following stats:
alter table tt1 update statistics set ('numRows'='10000000');
alter table tt2 update statistics set ('numRows'='10000000');
alter table tt3 update statistics set ('numRows'='2');
Could you add/modify the existing tests to set these stats and generate the MergeJoin-->MapJoin plan and verify that the fix works for that ?

Comment (ngsg) [reply ql/src/test/queries/clientpositive/antijoin2.q:20]: I updated the qfile and expected output for verifying MergeJoin -> MapJoin pattern.

However, I think this issue is unrelated with MergeJoin -> MapJoin pattern and depends only on MapJoinOperator. For example, the second query in this qfile contains only one Join, but the current Hive returns no rows although tt1 - tt2 is not empty. Also, I have checked that the current Hive returns wrong result for the first query without MergeJoin -> MapJoin pattern. (I used mvn test to run both of the queries.)

I think you could reproduce this issue with MapJoin -> MapJoin pattern if you disable vectorized execution. Could you please give me more details about the configuration that you used for reproducing the problem? Also I would be glad if you could run this qfile in your local environment and share the result.

Comment (scarlin-cloudera) [reply ql/src/test/queries/clientpositive/antijoin2.q:20]: We tested locally with the tpcds queries and your patch seems to have fixed the problem.

Thanks!

File: ql/src/java/org/apache/hadoop/hive/ql/exec/MapJoinOperator.java (hunk 1 above)
Comment (amansinha100) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/MapJoinOperator.java:559]: Just for reference, this check matches the check that is done in CommonJoinOperator.checkAndGenObject() here: https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/CommonJoinOperator.java#L973 .  Looks good to me.

**Review by amansinha100 — APPROVED** [review]
+1
Thanks for adding the additional tests.

**Comment by ramesh0201** [comment]
+1.  verified the test file with and without the patch. The patch fixes the issue.

**Review by kasakrisz — APPROVED** [review]
LGTM

**Pull request merged**